packages/apollo-server-express/src/expressApollo.ts

```diff
@@ -31,21 +32,52 @@ export function graphqlExpress(
     );
   }
 
-  return (req, res, next): void => {
+  const graphqlHandler = async (
+      req: express.Request,
+      res: express.Response,
+      next: express.NextFunction,
+  ) => {
     runHttpQuery([req, res], {
       method: req.method,
       options: options,
       query: req.method === 'POST' ? req.body : req.query,
       request: convertNodeHttpToRequest(req),
+      enableDefer: true,
     }).then(
-      ({ graphqlResponse, responseInit }) => {
+      async ({ graphqlResponse, graphqlResponses, responseInit }) => {
         if (responseInit.headers) {
-          for (const [name, value] of Object.entries(responseInit.headers)) {
-            res.setHeader(name, value);
-          }
+            for (const [name, value] of Object.entries(responseInit.headers)) {
+                res.setHeader(name, value);
+            }
+        }
+        if (graphqlResponse) {
```

Comment:
This check and a lot of its logic are being repeated for each integration (express, hapi, koa, etc). Did you consider abstracting some of it away for better future maintenance? Repeating the header, boundaries and the for loop seem like they could be in a stand-alone function.